aplib.net-demo/Assets/Scripts/Player/Movement.cs

```diff
@@ -0,0 +1,191 @@
+    }
+
+    /// <summary>
+    /// Draw a red wire sphere at the bottom of the player to visualize the ground check area.
+    /// </summary>
+    void OnDrawGizmosSelected() {
+        Gizmos.color = Color.red;
+        Gizmos.DrawWireSphere(_bottomPoint, _playerRadius);
+    }
+
+    /// <summary>
+    /// Main update loop where we do a ground check, move the player, and handle jumping.
+    /// Updates once per physics update.
+    /// </summary>
+    private void FixedUpdate() {
+        // Ground check by checking a sphere at the bottom of the player, more consistent than ray
+        _bottomPoint = transform.TransformPoint(_controller.center - Vector3.up * (_playerHeight * 0.55f - _playerRadius));
+        _isGrounded = Physics.CheckSphere(_bottomPoint, _playerRadius, _groundMask);
+
+        MovePlayer();
+        HandleJump();
+    }   
+    
+    /// <summary>
+    /// Move the player based on the input and the ground check using AddForce.
+    /// Also applies custom gravity and limits the player's velocity.
+    /// </summary>
+    private void MovePlayer() {
+        _horizontalVelocity = transform.right * _horizontalInput.x + transform.forward * _horizontalInput.y;
+        bool isOnSlope = OnWalkableSlope(out Vector3 directionOnSlope, out RaycastHit downHit);
+        bool isWallColliding = WalkingAgainstWall(out Vector3 wallNormal);
+        _rigidbody.useGravity = !isOnSlope;
+
+        if (_isGrounded) {
+            _rigidbody.drag = _groundDrag; 
+            
+            if (isOnSlope) {
+                _rigidbody.AddForce(_maxSpeed * _acceleration * Time.fixedDeltaTime * 1.5f * directionOnSlope);
+                
+                // Apply gravity but perpendicular to the slope, to prevent sliding
+                _rigidbody.velocity += (_actualGravityScale - 1) * _gravity * Time.fixedDeltaTime * downHit.normal;
+            }
+            else {
+                _rigidbody.AddForce(_maxSpeed * _acceleration * Time.fixedDeltaTime * _horizontalVelocity.normalized);
+            }   
+        }
+        else {
+            _rigidbody.drag = 0;  
+            _rigidbody.AddForce(_maxSpeed * _acceleration * Time.fixedDeltaTime * _airMovementScale * _horizontalVelocity.normalized);
+
+            // Custom gravity for player while falling
+            _actualGravityScale = _rigidbody.velocity.y < 0 ? _fallingGravityScale : _normalGravityScale;
+            _rigidbody.velocity += (_actualGravityScale - 1) * _gravity * Time.fixedDeltaTime * Vector3.up;
+        }
+
+        if (isWallColliding) HandleWallCollision(wallNormal);
+        LimitVelocity(isOnSlope);
+    }
+
+    /// <summary>
+    /// Add force along the wall, to prevent sticking.
+    /// </summary>
+    /// <param name="wallNormal">The normal of the wall the player is walking against</param>
+    private void HandleWallCollision(Vector3 wallNormal) {
+        float angleMultiplier = Mathf.Clamp01(1.2f - Vector3.Angle(_horizontalVelocity, -wallNormal) / 90);
```

Comment:
I assume the 90 is a trigonometry thing? If so, it can stay, if not, should it be defined?